models/resnet/resnet_impl.hpp

```diff
@@ -0,0 +1,181 @@
+ * terms of the 3-clause BSD license.  You should have received a copy of the
+ * 3-clause BSD license along with mlpack.  If not, see
+ * http://www.opensource.org/licenses/BSD-3-Clause for more information.
+ */
+#ifndef MODELS_MODELS_RESNET_RESNET_IMPL_HPP
+#define MODELS_MODELS_RESNET_RESNET_IMPL_HPP
+
+#include "resnet.hpp"
+
+namespace mlpack {
+namespace models {
+
+template<typename OutputLayerType, typename InitializationRuleType,
+    size_t ResNetVersion>
+ResNet<OutputLayerType, InitializationRuleType, ResNetVersion>::ResNet() :
+    inputChannel(0),
+    inputWidth(0),
+    inputHeight(0),
+    numClasses(0)
+{
+  // Nothing to do here.
+}
+
+template<typename OutputLayerType, typename InitializationRuleType,
+    size_t ResNetVersion>
+ResNet<OutputLayerType, InitializationRuleType, ResNetVersion>::ResNet(
+    const size_t inputChannel,
+    const size_t inputWidth,
+    const size_t inputHeight,
+    const bool includeTop,
+    const bool preTrained,
+    const size_t numClasses) :
+    ResNet<OutputLayerType, InitializationRuleType, ResNetVersion>(
+        std::tuple<size_t, size_t, size_t>(
+        inputChannel,
+        inputWidth,
+        inputHeight),
+        includeTop,
+        preTrained,
+        numClasses)
+{
+  // Nothing to do here.
+}
+
+template<typename OutputLayerType, typename InitializationRuleType,
+    size_t ResNetVersion>
+ResNet<OutputLayerType, InitializationRuleType, ResNetVersion>::ResNet(
+    std::tuple<size_t, size_t, size_t> inputShape,
+    const bool includeTop,
+    const bool preTrained,
+    const size_t numClasses) :
+    inputChannel(std::get<0>(inputShape)),
+    inputWidth(std::get<1>(inputShape)),
+    inputHeight(std::get<2>(inputShape)),
+    numClasses(numClasses)
+{
+  // Config for different Versions.
+  if (ResNetVersion == 18)
+  {
+    numBlockArray = {2, 1, 2, 2};
+    builderBlock = "basicblock";
+  }
+  else if (ResNetVersion == 34)
+  {
+    numBlockArray = {3, 4, 6, 3};
+    builderBlock = "basicblock";
+  }
+  else if (ResNetVersion == 50)
+  {
+    numBlockArray = {3, 4, 6, 3};
+    builderBlock = "bottleneck";
+  }
+  else if (ResNetVersion == 101)
+  {
+    numBlockArray = {3, 4, 23, 3};
+    builderBlock = "bottleneck";
+  }
+  else if (ResNetVersion == 152)
+  {
+    numBlockArray = {3, 8, 36, 3};
+    builderBlock = "bottleneck";
+  }
+  else
+  {
+    mlpack::Log::Fatal << "Incorrect ResNet version. Possible Values are: 18, "
+        "34, 50, 101 and 152" << std::endl;
+  }
+
+  resNet.Add(new ann::Convolution<>(3, 64, 7, 7, 2, 2, 3, 3, inputWidth,
+      inputHeight));
+
+  // Updating input dimesntions.
+  inputWidth = ConvOutSize(inputWidth, 7, 2, 3);
+  inputHeight = ConvOutSize(inputHeight, 7, 2, 3);
+
+  mlpack::Log::Info << "Convolution: " << "(" << 3 << " " << 64 << " " << 7
+      << " " << 7 << " " << 2 << " " << 2 << " " << 3 << " " << 3 << " " <<
+      inputWidth << " " << inputHeight << ")" << std::endl;
+
```

Comment:
Printing out Params like this doesn't make much sense. See the below image for Darknet. Also added an image for model.summary in PyTorch. A good user-experience would be simplifying things.
<img width="428" alt="Screenshot 2021-06-19 at 6 29 32 PM" src="https://user-images.githubusercontent.com/39593019/122643247-65b41900-d12c-11eb-998e-6405ab019820.png">
<img width="562" alt="27968068-89234e14-6313-11e7-92dc-92c31b8e28b9" src="https://user-images.githubusercontent.com/39593019/122643305-b0359580-d12c-11eb-90d7-1a65e2897a59.png">

Reply:
Makes sense, btw pytorch does throws out all the parameters that you have in your model, i think this example is from tensorflow, but i do like the ones pytorch has so that made me implement all the params

Reply:
Ahh, its from Keras I believe. Its much cleaner. The user only cares about the dim changes since if he is going to use a pre-trained model he doesn't care about layer params since he won't be changing them.

Reply:
Will make this change, btw how do you see the otuput of `mlpack::Log::Info`